pkg/pxc/app/statefulset/haproxy.go

```diff
@@ -0,0 +1,218 @@
+package statefulset
+
+import (
+	"fmt"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	app "github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/app"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	haproxyName           = "haproxy"
+	haproxyDataVolumeName = "haproxydata"
+)
+
+type HAProxy struct {
+	sfs     *appsv1.StatefulSet
+	labels  map[string]string
+	service string
+}
+
+func NewHAProxy(cr *api.PerconaXtraDBCluster) *HAProxy {
+	sfs := &appsv1.StatefulSet{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "apps/v1",
+			Kind:       "StatefulSet",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-" + haproxyName,
+			Namespace: cr.Namespace,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			PodManagementPolicy: "OrderedReady",
+		},
+	}
+
+	labels := map[string]string{
+		"app.kubernetes.io/name":       "percona-xtradb-cluster",
+		"app.kubernetes.io/instance":   cr.Name,
+		"app.kubernetes.io/component":  haproxyName,
+		"app.kubernetes.io/managed-by": "percona-xtradb-cluster-operator",
+		"app.kubernetes.io/part-of":    "percona-xtradb-cluster",
+	}
+
+	return &HAProxy{
+		sfs:     sfs,
+		labels:  labels,
+		service: cr.Name + "-" + haproxyName,
+	}
+}
+
+func (c *HAProxy) AppContainer(spec *api.PodSpec, secrets string, cr *api.PerconaXtraDBCluster) (corev1.Container, error) {
+	initDelay := int32(10)
+	if spec.LivenessInitialDelaySeconds != nil {
+		initDelay = *spec.LivenessInitialDelaySeconds
+	}
+	appc := corev1.Container{
+		Name:            haproxyName,
+		Image:           spec.Image,
+		ImagePullPolicy: corev1.PullAlways,
+		Ports: []corev1.ContainerPort{
+			{
+				ContainerPort: 3306,
+				Name:          "mysql",
+			},
+			{
+				ContainerPort: 1024,
+				Name:          "stat",
+			},
+		},
+		TerminationMessagePath: "/dev/termination-log",
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      "haproxy-cfg",
+				MountPath: "/etc/haproxy/pxc",
+			},
+			{
+				Name:      "haproxy-auto",
+				MountPath: "/etc/haproxy-auto/",
+			},
+		},
+		Env: []corev1.EnvVar{
+			{
+				Name:  "POD_NAME",
+				Value: c.sfs.ObjectMeta.Name,
+			},
+			{
+				Name:  "POD_NAMESPACE",
+				Value: c.sfs.ObjectMeta.Namespace,
+			},
```

Comment:
@hors why `POD_NAME` and `POD_NAMESPACE` are needed?

Reply:
I do not have such env vars in statefulset.

Reply:
I do not have such env vars in statefulset.